### ISSUE TITLE
Add arrow operator as sugar for (*var).

### ISSUE THREAD

**Issue by wrightwriter** [gpuweb/gpuweb#4114]
```rust
val b = (*f).bar;
val f = &(*f).bar;
```
turns into
```rust
val b = f->bar;
val f = &f->bar;
```
This would make pointers easier to write and more readable. Would also make WGSL more accessible to people without systems language experience.

**Comment by litherum** [comment]
Should this be retitled “Add arrow operator”? Or maybe “Add arrow operator as sugar for *.”?

**Comment by ben-clayton** [comment]
One alternative is to allow `.` on pointers, resulting in a pointer to that member.

So:

```wgsl
val b = (*p).bar;
let f = &(*p).bar;
```

Would be equivalent to:

```wgsl
val b = *p.bar;
let f = p.bar;
```

**Comment by kainino0x** [comment]
Or to allow `.` on pointers, resulting in a reference to that member, which can be turned into a pointer by `&`.

**Comment by ben-clayton** [comment]
> Or to allow `.` on pointers, resulting in a _reference_ to that member, which can be turned into a pointer by `&`.

[That's actually what golang does](https://go.dev/play/p/E3jZiy874cv) - so +1 from me. :-)

**Comment by ben-clayton** [comment]
>  `.` produces a reference, this effectively hides away a potentially slow read/write from the reader of the code.

I don't quite follow the argument here.
A reference in WGSL is essentially a pointer. The load only happens when invoking [the load rule](https://www.w3.org/TR/WGSL/#load-rule). So, using `.` or `->` doesn't change the performance of the expression.

**Comment by wrightwriter** [comment]
You keep implying I am suggesting a difference in performance, because of the syntax. I am not.
   
You don't see why global reads/writes might matter to a reader of the shader?     
Well I think being explicit about where you are writing to a `ptr<function>` variable is also super useful.
Both of these are issues which can be seen in GLSL's inout and buffer syntax. WGSL's `(*ptr)` is an improvement over this.

Also, most implementations substituting the variable is unrelated to you being able to see where this is happening.
This point also misses the fact that you can have pointers to values which can't get stored in a variable, like unsized arrays or an atomic or a struct which contains them. Also, you do not have to read the value behind the pointer at all. You can use the pointer as a proxy for getting another pointer to, say, an element of an array or a member of a struct.   

Anyways, here's the example rewritten with a valid (simplified) case which I encounter often and which demonstrates the readability point again:
```rust
fn F() { 
  let p: ptr<storage,Particle,read_write> = &some_storage_buffer[4];
  // a lot of code here
  p.pos += 1; // the global access is hidden
  p->pos += 1; // easy to see
  *p.pos += 1; // easy to see
  (*p).pos += 1; // easy to see
  // a lot of code here
}
```

**Comment by ben-clayton** [comment]
> You keep implying I am suggesting a difference in performance, because of the syntax. I am not.

Sorry, I'm just really struggling to understand your argument. I assumed your argument was focused around performance as you said:

> Explicitness matters here, especially in performance critical code.

> You don't see why global reads/writes might matter to a reader of the shader?

I'm not sure what you mean by `global` here. Do you mean storage buffers? Values that aren't in registers? Something else?

The point that I'm trying to make is that `->` does not help the reader know what address space the LHS memory is in. You can have pointers to `function`, `private`, `uniform`, `storage`, `workgroup`.
Likewise, you can currently use `.` for `var`s in all these address spaces too.
`function` and `private` are typically GPU registers, so a `ptr` does not imply that it's accessing some non-register form of memory.

> I think being explicit about where you are writing to a ptr<function> variable is also super useful.

So this is possibly the part we're disagreeing on.

If I understand correctly, your argument is that it's super useful to see **on the line that you perform an access** that the left-hand symbol resolves to a `ptr` instead of a non-ptr `let` or `var`.

Given that a `ptr` is just an alias to some other variable's storage, I'm still unconvinced by the value in making a `ptr` access syntax special.
There's an argument that the aliasing itself is a hazard, but WGSL has [fairly strict limitations on what can be aliased](https://www.w3.org/TR/WGSL/#alias-analysis).

> Anyways, here's the example rewritten with a valid (simplified) case which I encounter often and which demonstrates the readability point again:

Agree that it's easy to distinguish between accesses on `ptr` vs non-`ptr`. Disagree that this is actually valuable to distinguish.

**Comment by litherum** [comment]
Many pointers in WGSL will be compiled away to nothing (no memory access at all). All pointers to local variables will be compiled away to nothing. So I'm not sure telling authors "beware, whenever you type the `->` or `*` operator, your code is going to be slow" is the right message. I think the right message is "when you see a `->` or a `*` operator, you have to look at what is being dereferenced to determine the performance characteristics of that access."

**Comment by litherum** [comment]
If we make the `.` operator valid on pointers, that would mean we can't give real methods to pointers in the future.

For example, if we add a lambda feature, I could totally imagine a future where you could write this kind of thing:
```
let p: ptr<int, function> = ...;
p.runWithMemoryRebound<f32>([] (reboundPointer: ptr<f32, function>) {
    ... *reboundPointer ...
});
```

**Comment by ben-clayton** [comment]
> If we make the `.` operator valid on pointers, that would mean we can't give real methods to pointers in the future.

True, but that's pretty unusual to put methods on a pointer type - do you have any example languages that do this?
Some might argue that it would be confusing to have `.` methods for the pointer and `->` or `(*p).` for the pointee.

> For example, if we add a lambda feature, I could totally imagine a future where you could write this kind of thing:

Sure, but you could just as easily pass the pointer as the first argument to a free function:

```
let p: ptr<int, function> = ...;
runWithMemoryRebound<f32>(p, [] (reboundPointer: ptr<f32, function>) {
    ... *reboundPointer ...
});
```

That is more fitting with our existing builtins (e.g. `bitcast`)

**Comment by ben-clayton** [comment]
> I think the right message is "when you see a `->` or a `*` operator, you have to look at what is being dereferenced to determine the performance characteristics of that access.

Or a `.`. As I keep saying: references (the type of an identifier resolving to a `var`) have the exact same address spaces as `ptr`. There's simply nothing special about `ptr` accesses.

**Comment by dneto0** [comment]
I agree WGSL should add a convenience for this.
This became super clear when writing a tutorial example of using pointers as  short names within a function.  https://google.github.io/tour-of-wgsl/types/pointers/pointers_as_short_names/ 
```
struct Particle {
   momentum: vec3f,
   pos: vec3f,
   velocity: vec3f,
}
@group(0) @binding(0)
var<storage,read_write> particles: array<Particle>;

fn sequentially_update_momenta() {
  let particles_ptr = &particles;  // Here is a pointer
  for (var i: u32; i < arrayLength(particles_ptr); i++) {
    // Use & to get a pointer as a short name for something inside a big variable.
    let a_particle = &particles[i];

    // We could have spelled out the pointer type in full.
    // let a_particle: ptr<storage,Particle,read_write> = &particles[i];

    // Now use 'a_particle' as a ahort name.
    let its_pos = (*a_particle).pos;  // Parentheses are required.
    let its_velocity = (*a_particle).velocity;
    (*a_particle).momentum = its_pos * its_velocity;
  }
}
```
Those last three code lines feel awfully wordy for an idiomatic use.

**Comment by alan-baker** [comment]
As I mentioned in the 2023-05-30 meeting, having `.` dereference doesn't feel great from a philosophical standpoint for me. I think there is a benefit on the cognitive side to encourage users to leave things in the pointer domain as long as possible. Loads and stores are comparatively expensive on device and I think there is benefit in trying to expose that fact in the language itself. So to me, #1580 is preferable to dereferencing via `.`.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-05-30-minutes#add-dereference-operator--issue-4114">WGSL 2023-05-30 Minutes</a></summary>

* JB: Both Go and Rust and Java treat . as dereference-then-get-member. That would be totally fine.
* BC: “Dereference” doesn’t perform the memory access. Changes from ptr to ref.  
* JB: So the foo.blah has type reference-to-the-blah-type.
* DN: Fine with me.  Should do something in this space, experience after writing samples.
* BC: Alan said this lessens the “need” for reference type, it seems.
* AB: One difference would be, if you called variable pointer by default, then you don’t always need a star to dereference it.  I’m on the fence as to whether this is a reference or not.
* MM: Have concern overloading dot operator.  Not aware of languages that have both pointers and non-pointers.  Don’t have a pointer type by itself.
* JB: C# does that.
* BC: Go has pointer types, almost same syntax as C.  And dot yields what we’re discussing.
* JB: Think Go, Rust and C# are the good analogies. They handle this just fine.
* MM: Swift has function on a pointer type of (?) and memory rebound.
* JB: Rust does this a lot.  When dealing with pointers it’s so commonly done it’s reserved for doing the dereference.  If you want a function to acton a pointer, it has to be a free function, not a member function ont he pointer type.
* KG: Sounds rough consensus for doing the . as described?
* DN: “language extension”.
* BC: Part of a v2?
* AB: Well, [worth considering that the] main use we have for pointers is to create a short name. This won’t work for that; still need to throw in an ampersand. Having it return a pointer, would obviate needing to put in a & to get the pointer back. Seems like a too-quick decision
* JB: Not clear about the cases you’re concerned about. Please put them into the issue.
* MM: sounds like if I want to make a short name.
    * A.b.c.d.e; can make a short name for that. But then if I used a pointer ot make a short name, I still need to use & and *.  But sounds like you’re still wanting to make C++ references.
* Alan Baker
    * let a = &x.y.z;  // a is  pointer.
    * let b = a.c;  // To make  ‘b’ a pointer, would need to throw in an & to get a pointer back.
* JB: In C++, if you use arrow operator, you’d have to use & to get it back to a pointer.
* KG: This reads naturally. B is a value, not a pointer.
* MM: in this example, to make a value,   let b = a->c;  Type of b is value.
* BC: Existing Go
    * [https://go.dev/play/p/Uq5SC1LioPX](https://go.dev/play/p/Uq5SC1LioPX) 
    * // You can edit this code!
    * // Click here and start typing.
    * package main
    * 
    * import "fmt"
    * 
    * type C struct {
    * 	i int
    * }
    * 
    * type B struct {
    * 	c C
    * }
    * type A struct {
    * 	b B
    * }
    * 
    * func main() {
    * 	a := A{}
    * 
    * 	p_a := &a  // a pointer value.
    * 
    * 	p_b := &p_a.b   // yup, this is what Alan is talking about.  Is a pointer value.
    * 
    * 	fmt.Println(a, p_a, p_b)
    * 
    * }
    * 
* AB: Some time ago David proposed a symmetry of ptr   . and [] to yield pointers.  I personally prefer that over this.  That’s closer to what I’m saying now.  I like the symmetry, but can read the room that I’m in the minority.
* BC: It’s going against the trend of existing languages, and doesn’t address the change of having to put * all over the place.
* ..
* AB: I’d like to encourage people to stay in pointer land, to delay the memory access. It’s cheaper generally to stay operating on addresses.  You normally rely on on underlying compilers to clean it up for you. Think it’s beneficial to encourage that in the source language.
* JB: Please put examples in the issue.
* MM: **Think Alan’s argument is worthy writing in the issue.**
* AB: Will add it.

</details>

**Comment by jimblandy** [comment]
I see the analogy with `OpAccessChain` and LLVM's `getelementptr`. But:

1) There is no other language in which `.member` produces a pointer.

2) Exactly where loads and stores appear is a little weird in *all* Algol-derived languages - anything with lvalues and rvalues, `ref` and `ptr`, etc.

**Comment by kainino0x** [comment]
> > do you have any example languages that do this?
> 
> Swift.

Come to think of it, I think Rust has it too, in the form of `impl SomeTrait for &SomeType`. But if `SomeType` has a method named `foo`, I think it shadows the `foo` on `&SomeType`.

Anyway, that to say, I don't see any reason you can't have both methods on pointers and `.member` on pointers. You just have to write the resolution rules such that collisions cause either an error or shadowing.

**Comment by litherum** [comment]
If I understand correctly, there are 3 proposals at hand:

1. Match C: the lhs of a `.` must not be a pointer. Add a new operator, `->`, for which `foo->bar` is shorthand for `(*foo).bar`.
2. Match Go: Don't add a new operator. The lhs of `.` may be a pointer or a non-pointer. If `foo` is a pointer, then `foo.bar` is a shorthand for `(*foo).bar`. Otherwise, the `.` works just as it does today.
3. Symmetry: Don't add a new operator. The lhs of `.` may be a pointer or non-pointer. if `foo` is a pointer, then `foo.bar` is a shorthand for `&((*foo).bar)`. Otherwise, the `.` works just as it does today.

(I'm not really counting the proposals like `foo +. bar` though I kind of like them personally, I agree that they're too wacky to be considered realistic.)

Here are some examples:

## Example 1 "Double pointer"

```
let p : ptr<function, ptr<function, i32>> = ...;
let i = /* dereference p twice */;
```

Here, all proposals are the same:
```
let i = **p;
```

## Example 2 "Simple uniform"

```
struct Foo {
    bar : i32,
}
let p : ptr<uniform, Foo> = ...;
/* assign to the bar field */
```

Proposal 1:
```
p->bar = 42;
```
Proposal 2:
```
p.bar = 42;
```
Proposal 3:
```
*p.bar = 42;
```

## Example 3 "Advanced Uniform"

```
struct Foo {
    bar : i32,
}
struct Baz {
    foo : Foo,
}
let p : ptr<uniform, Baz> = ...;
/* assign to the inner bar field */
```

Proposal 1:
```
p->foo.bar = 42;
```
Proposal 2:
```
p.foo.bar = 42;
```
Proposal 3:
```
*p.foo.bar = 42;
```

## Example 4 "Linked List"

```
struct Foo {
    bar : i32,
}
struct Baz {
    foo : ptr<function, Foo>,
}
let p : ptr<function, Baz> = ...;
/* assign to the inner bar field */
```

Proposal 1:
```
p->foo->bar = 42;
```
Proposal 2:
```
p.foo.bar = 42;
```
Proposal 3:
```
*p.foo.bar = 42;
```

## Example 5 "Dynamic Resource Index"

```
struct Foo {
    bar : i32,
}
struct Baz {
    foo : ptr<function, Foo>,
}
let p : Baz= ...;
/* assign to the inner bar field */
```

Proposal 1:
```
p.foo->bar = 42;
```
Proposal 2:
```
p.foo.bar = 42;
```
Proposal 3:
```
*p.foo.bar = 42;
```

## Example 6 "Crazytown"

```
struct Foo {
    bar : i32,
}
struct Baz {
    foo : ptr<function, ptr<function, Foo>>,
}
let p : Baz= ...;
/* assign to the inner bar field */
```

Proposal 1:
```
(*p.foo)->bar = 42;
```
Proposal 2:
```
(*p.foo).bar = 42;
```
Proposal 3:
```
*(*p.foo).bar = 42;
```

**Comment by litherum** [comment]
Looking at the above, we don't really have much preference among options 2 and 3, but we mostly just dislike option 3 because of all the `*`s you need to write to use it. That being said, we realize that this is mostly a stylistic decision.

**Comment by jimblandy** [comment]
@litherum This is a *superb* summary. By "we", you mean Apple, right?

**Comment by jimblandy** [comment]
Example 4, proposal 3, seems incorrect.

The fully parenthesized form of proposal 3 is `(*((p.foo).bar)) = 42;`

Leaving out the address spaces, the type of `p.foo` is `ptr<ptr<Foo>>`.

Per the proposal, `.` only works on structs and pointers to structs - it doesn't follow an arbitrary number of pointers. So `(p.foo).bar` is a type error.

Instead, the proposal 3 version must be `*(*p.foo).bar = 42;`.

**Comment by kdashg** [comment]
I expected Proposal 2 to be recursive, such that p:Foo**`, the compiler says:
> * Oh, `p` in `p.foo` is a ptr, so I'll emit `(*p).foo)`.
> * Oh, `(*p)` in `(*p).foo` is a ptr, I'll emit `(*(*p)).foo`.
> * Yes, `(*(*p))` has a `.foo`, great!

**Comment by jimblandy** [comment]
Rust's `.` operator does follow as many levels of pointers as you have. If you've got `p` of type `&&&&&&&&&&&&&&&Foo`, then `p.x` works.

Go does not do this. The Go `.` operator follows only one level of pointers.  https://go.dev/ref/spec#Selectors

(I heard this point go by in the committee, but I didn't speak up because I felt like the conversation was probably too hard to follow already. Apologies for that.)

**Comment by dneto0** [comment]
> Despite that, I'd been assuming that most people were assuming we were only considering at most one level of indirection. I'm not so sure any more, so it's worth clarifying.

Right.  I have been assuming only one level of indirection is followed.  

Following "as many as it takes" strikes me as bad because  (I think) that means doing a possibly unbounded number of accesses (?), but that "loop" is impossible to to "see" in the code.

**Comment by litherum** [comment]
Today on the call we resolved on option 2.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-06-13-minutes#add-dereference-operator--issue-4114">WGSL 2023-06-13 Minutes</a></summary>

* AB: Examples 4 and 5 look the same to me.  Also, there’s nothing with array acceses; which is what I think of for “dynamic” for me.
* KG: Solutions for 4 proposal 3 had typo. 
* KG: I expected proposal 2 to be recursive. But it is not, per Go rules.  
* MM: With null proposal, are there only 4.
* DN: Think any extra would be in no existing language. Let’s not go there.
* MM: So with 4 proposals.  Think it’s mostly style.  Our main feeling is option 3 is worse.
* JB: Want option 2.
* BC: Want option 2.
* DN: Symmetry option also tells you what to do for p[i].   Can extend option2 to be similar: p[i] would yield reference.
* DN: I’m ok with option 2.  Improves over 3 because 3 has the problem of saying * on one end of expression and . on the other end. Makes it harder to read.
* JB: Pascal solved this by having post-fix dereference.
* KG: Most common thing you want to do with this is to make alias/shorthands. 
* MM: Common case is to make an alias to a field of a resource. Will need to write & anyway.
* KG: Right have to start 
* **KG: Consensus for option 2.**  Despite the coolness of the symmetry of 3.
* AB: Have slight preference to 1 because of familiarity.  But not a big deal.
* **KG: Consensus for option 2 with additional sugar for array indices, and Go-style of only dereference on level right now. (We don’t have ptr-to-ptr right now so the last point makes no difference)**

</details>

**Comment by lain-dono** [comment]
A special dereferencing syntax is not useful. All necessary information is contained in the variable type. You see a pointer - you know about dereferencing. All I can do with a pointer is dereference it.

Besides, a simpler syntax is better. Easier to parse and easier to generate. And of course it's easier to read, too.
